libs/agno/agno/app/agui/async_router.py

```diff
@@ -0,0 +1,366 @@
+        yield RunStartedEvent(type=EventType.RUN_STARTED, thread_id=run_input.thread_id, run_id=run_id)
+
+        # Request streaming response from agent
+        response_stream = await agent.arun(
+            message=message,
+            session_id=run_input.thread_id,
+            stream=True,
+            stream_intermediate_steps=True,
+        )
+
+        # Stream the response content
+        async for event in _stream_response_content(
+            response_stream=response_stream, thread_id=run_input.thread_id, run_id=run_id
+        ):
+            yield event
+
+    # Emit a RunErrorEvent if any error occurs
+    except Exception as e:
+        logger.error(f"Error running agent: {e}", exc_info=True)
+        yield RunErrorEvent(type=EventType.RUN_ERROR, message=str(e))
+
+
+async def run_team(team: Team, input: RunAgentInput) -> AsyncIterator[BaseEvent]:
+    """Run the contextual Team, mapping AG-UI input messages to Agno format, and streaming the response in AG-UI format."""
+    run_id = input.run_id or str(uuid.uuid4())
+    try:
+        # Extract the last user message for team execution
+        user_message = _get_last_user_message(input.messages) if input.messages else ""
+        yield RunStartedEvent(type=EventType.RUN_STARTED, thread_id=input.thread_id, run_id=run_id)
+
+        # Request streaming response from team
+        response_stream = await team.arun(
+            message=user_message,
+            session_id=input.thread_id,
+            stream=True,
+            stream_intermediate_steps=True,
+        )
+
+        # Stream the response content
+        async for event in _stream_team_response_content(
+            response_stream=response_stream, thread_id=input.thread_id, run_id=run_id
+        ):
+            yield event
+
+    except Exception as e:
+        logger.error("Error in run_team: %s", e, exc_info=True)
+        yield RunErrorEvent(type=EventType.RUN_ERROR, message=str(e))
+
+
+async def _stream_response_content(
+    response_stream: AsyncIterator[RunResponse], thread_id: str, run_id: str
+) -> AsyncIterator[BaseEvent]:
+    """Map the Agno response stream to AG-UI format."""
+    message_id = str(uuid.uuid4())
+    message_started = False
+    active_tool_call_id = None
+    ended_tool_call_ids = []
+
+    async for chunk in response_stream:
+        content = _extract_response_chunk_content(chunk)
+
+        # Handle text responses
+        if chunk.event == RunEvent.run_response:
+            # Emit an event fTeamRunResponse the message
+            if not message_started:
+                message_started = True
+                yield TextMessageStartEvent(
+                    type=EventType.TEXT_MESSAGE_START,
+                    message_id=message_id,
+                    role="assistant",
+                )
+            # We need to end any active tool call if it exists and it hasn't been ended yet
+            if active_tool_call_id is not None:
+                yield ToolCallEndEvent(
```

Comment:
We can only do this if we got the completed event though. Don't we need to give the result of the tool call?

Reply:
We don't seem to present the result anywhere indeed. The ToolCall events are the start, args and end ones, and the end one just receives the tool call id?: https://docs.ag-ui.com/concepts/events#toolcallend

I may be missing something here